### PR TITLE
Feature/segmentation colors

### PR DESCRIPTION
**Pull request by toloudis**
This adds a function to generate a volume intensity lookup table that maps each different intensity value to a unique color.
This would be used to automatically color a labeled image (in which the labels are distinct discrete scalar intensities in the volume).

The algorithm for choosing random colors is not yet perfect for this and may be refined in a future PR.

There will be PRs associated with this change to expose it in the ui (allen-cell-animated/website-3d-cell-viewer) and in jupyter notebooks (allen-cell-animated/nbvv).  volume-viewer is the lowest level code that requires the change.

### CODE REVIEW

**Comment by toloudis** [comment]
**The important changes are in Histogram.js.**  Most of the remaining changes are due to formatting and updating three.js.

File: src/Histogram.js [hunk 1]
```diff
@@ -159,6 +161,61 @@ export default class Histogram {
...
   }
 
+  /**
+   * Generate a lookup table with a different color per intensity value
+   * @return {Lut}
+   */
+  lutGenerator_labelColors() {
+    const lut = new Uint8Array(256 * 4).fill(0);
+    const controlPoints = [];
+    controlPoints.push({ x: 0, opacity: 0, color: [0, 0, 0] });
```
[the remark this answers is not in the document]
Comment (toloudis) [reply src/Histogram.js:171]: The control points are for plotting the data in the transfer function editor.  Ideally we can generate the LUT from the control points, and the control points should be the true source of data but that's a bigger change that probably breaks public api also.  When the interactive TF editor is being used, it does that LUT generation given the set of user edited control points.

File: src/Histogram.js [hunk 2]
```diff
@@ -159,6 +161,61 @@ export default class Histogram {
...
+        a = 1;
+      } else {
+        // add a zero control point?
+        r = 0;
+        g = 0;
+        b = 0;
+        a = 0;
+      }
+      // if current control point is same as last one don't add it
+      if (r !== lastr || g !== lastg || b !== lastb || a !== lasta) {
```
Comment (schoinh) [inline src/Histogram.js:204]: Does this mean that the values in `this.bin` are sorted? i.e. we don't need to compare the rgba values with any other past values than the last set of values?

File: src/Histogram.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (toloudis) [reply src/Histogram.js:171]: AICS 3D Cell Viewer 2021-06-09 13-34-03

File: src/Histogram.js (hunk 2 above)
Comment (toloudis) [reply src/Histogram.js:204]: if there are a bunch of empty bins in between bins with data, we don't want to keep inserting the same control point over and over.  Basically we only want to add a control point if its information would be DIFFERENT than the previous one.  To keep the number of control points as low as possible.  The control points get connected linearly.  Which is now making me think that this algorithm is going to have some errors. I really need to make a unit test for this control point + LUT stuff.

File: src/Histogram.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (schoinh) [reply src/Histogram.js:171]: Is `opacity` here what's plotted on the vertical axis here?

File: src/Histogram.js [hunk 3]
```diff
@@ -308,44 +368,32 @@ export default class Histogram {
...
-          });
-          break;
+        lastOpacity = opacity;
+        opacity = Math.clamp(Math.round(255 * (map[i] - map[0])), 0, 255);
+        lut[i * 4 + 3] = opacity;
+
+        slope = opacity - lastOpacity;
+        // if map[i]-map[i-1] is the same as map[i+1]-map[i] then we are in a linear segment and do not need a new control point
+        if (slope != lastSlope) {
+          lutControlPoints.push({ x: i - 1, opacity: lastOpacity / 255.0, color: [255, 255, 255] });
```
Comment (schoinh) [inline src/Histogram.js:392]: How come the color is always 255, 255, 255 here? Is that what "equalized" lookup table means?

**Review by schoinh — APPROVED** [review]
There's a lot I still don't understand, but I trust that this PR is beneficial and functional based on your demo and the new tests :D

**Review by meganrm — APPROVED** [review]

**Comment by toloudis** [comment]
Thanks for the reviews.  After a UX meeting with Graham and Thao, we are going to make some revisions to how the color mapping is applied.  I'll merge this and then do a followup PR with the new functionality.  Summary:  the colorize LUT is going to alpha-blend (weighted multiply) against the existing LUT, rather than completely replacing it.

File: src/Histogram.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (toloudis) [reply src/Histogram.js:171]: yes

File: src/Histogram.js [hunk 4]
```diff
@@ -202,6 +202,9 @@ export default class Histogram {
       }
       // if current control point is same as last one don't add it
       if (r !== lastr || g !== lastg || b !== lastb || a !== lasta) {
+        if (lasta === 0) {
+          controlPoints.push({ x: i - 0.5, opacity: lasta, color: [lastr, lastg, lastb] });
+        }
```
[the remark this answers is not in the document]
Comment (toloudis) [reply src/Histogram.js:207]: > There's a lot I still don't understand, but I trust that this PR is beneficial and functional based on your demo and the new tests :D

I can do a longer deep dive with you on how these lookup tables are used, what they do and why.  It's really about controlling how the volume data appears on screen -- being able to pick out signal from noise in the data.

File: src/Histogram.js (hunk 3 above)
Comment (toloudis) [reply src/Histogram.js:392]: The preset functions that generate lookup tables don't know anything about color.  They all use 255,255,255 for the color (except for the brand new colorize one) and only really modulate opacity.  To be honest, the Equalize function doesn't get used by any client code right now but it was one where the LUT and the control points were incompatible (based on unit tests!)  So I put in this change to fix that.

**Pull request merged**